File: grind75/Dec28/322.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        #knapsack, inifinite amount of each item
        dp = [math.inf] * (amount+1)
        dp[0] = 0
        
        for coin in coins:
            for i in range(1, amount + 1):
                if i - coin >= 0:
                    dp[i] = min(dp[i], dp[i-coin] + 1)
        
        if dp[-1] == math.inf:
            return -1
        return dp[-1]

File: grind75/Dec28/322.go
func coinChange(coins []int, amount int) int {
    q := []int{0}
    visited := map[int]struct{}{}
    visited[0] = struct{}{}
    sort.Ints(coins)
    cnt := 0
    for len(q) > 0 {
        cq := []int{}
        lq := len(q)
        for lq > 0 {
            cur := q[0]
            q = q[1:]
            lq--
            if cur == amount {
                return cnt
            }
            for _, val := range coins {
                if val+cur > amount {
                    break
                }
                if _, ok := visited[val+cur]; !ok {
                    visited[val+cur] = struct{}{}
                    cq = append(cq, val+cur)
                }
            }
        }
        q = cq
        cnt += 1
    }
    return -1
}
